File: Forensics/Android Forensics/Tools/gesture_cracker.c
/*
gcc -o gesture_cracker gesture_cracker.c -lcrypto
*/

#include <openssl/sha.h>
#include <stdio.h>
#include <string.h>

unsigned char hash[SHA_DIGEST_LENGTH];


void try_permutations(unsigned char target_hash[], char arr[], int n, int l, int index, char data[])
{
    if (index == l) {

        // Try the current permutation
        SHA1(data, l, hash);

        // Check if the hash matches
        int match = 1;

        for (int i = 0; i < SHA_DIGEST_LENGTH; i++) {
            if (hash[i] != target_hash[i]) {
                match = 0;
                break;
            }
        }

        // Print the result
        if (match) {
            printf("Found a match: ");
            for (int i = 0; i < 9; i++) {
                printf("%c", data[i] + '0');
            }
            printf("\n");
        }
    }
    else {
        // Compute the list of used values in data[0:index]
        char used[n];
        memset(used, 0, n);
        for (int i = 0; i < index; i++) {
            used[data[i]] = 1;
        }

        // Try all possible values after the current index
        for (int i = 0; i < n; i++) {
            if (!used[i]) {
                data[index] = arr[i];
                try_permutations(target_hash, arr, n, l, index + 1, data);
            }
        }
    }
}

int main(int argc, char const *argv[]) {

    // Check for argument
    if (argc < 2) {
        printf("Usage: %s <hash>\n", argv[0]);
        printf("The hash is the content of android/data/system/gesture.key as a string of hex values\n");
        return 1;
    }

    // Read the argument to a buffer
    int gesture_len = strlen(argv[1]);

    // Convert the hex string to bytes
    unsigned char target_hash[gesture_len / 2];
    for (int i = 0; i < gesture_len; i += 2) {
        sscanf(argv[1] + i, "%2hhx", &target_hash[i / 2]);
    }

    // Print the input hash
    printf("Input hash: ");
    for (int i = 0; i < SHA_DIGEST_LENGTH; i++) {
        printf("%02x", target_hash[i]);
    }
    printf("\n");

    // Define the possible values
    char gesture_chars[] = "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09";

    // Iterate over all possible lengths
    for (int l = 2; l <10; l++) {

        // Iterate over all possible permutations of gesture_chars
        char data[l];
        try_permutations(target_hash, gesture_chars, 10, l, 0, data);
    }
    return 0;
}

File: Forensics/Android Forensics/Tools/gesture_cracker.py
import itertools
import sys
from hashlib import sha1

def sha1_crack(target_hash:str):
    """
    Cracks the SHA1 hash of a gesture.key file and prints the gesture as a string of integers

    Args:
        target_hash (str): The SHA1 hash of the gesture.key file in hexadecimal
    """

    # List of all possible characters
    gesture_chars = ["\x00","\x01","\x02","\x03","\x04","\x05","\x06","\x07","\x08","\x09"]
    print ("hash = {}".format(target_hash))

    # Iterate through all possible lengths of the gesture
    for i in range(2, 10):

        # Iterate through all possible permutations of the gesture with the given length
        permutations = [(''.join(p)).encode('utf-8') for p in itertools.permutations(gesture_chars, i)]
        for j in permutations:
            current_hash = sha1(j).hexdigest()
            if target_hash == current_hash:

                # Print the gesture as a string of integers
                print ("gesture = {}".format(j.hex()[1::2]))

if __name__ == "__main__":

    # Parse user input
    if (len(sys.argv) < 2):
        print ("Usage: python3 gesture_cracker.py <hash in hex format>")
        print ("   or: python3 gesture_cracker.py <path to gesture.key>")
        exit()
    param1 = sys.argv[1]

    # If the parameter is a path, read the file
    if ("/" in param1):
        try:
            with open(param1, "rb") as f:
                hash = f.read().hex()
        except:
            print ("Error: " + param1 + " is not a valid file, try passing the hash directly")
            exit()
    else:
        hash = param1
    sha1_crack(hash)